File: Data Structures & Algorithms/house-robber-ii/submission-3.java
class Solution {
    public int rob(int[] nums) {
        return Math.max(rec(nums, 0, nums.length - 1), rec(nums, 1, nums.length));
    }
    public int rec(int[] nums, int i, int end) {
        if (i >= end) {
            return 0;
        }
        return Math.max(nums[i] + rec(nums, i + 2, end), rec(nums, i + 1, end));
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        mini = nums[0]
        maxi = nums[0]
        res = nums[0]
        for i in range(1, len(nums)):
            x = nums[i]
            maxi = max(x, maxi * nums[i], mini * nums[i])
            mini = min(x, mini * nums[i], maxi * nums[i])
            res = max(res, maxi, mini)
        return res
